app/controllers/registrations_controller.rb

```diff
@@ -0,0 +1,57 @@
+# frozen_string_literal: true
+
+class RegistrationsController < ApplicationController
+  def new
+  end
+
+  def create
+    user = User.new(username: session_params[:username])
+
+    credential_options = WebAuthn.credential_creation_options
+    credential_options[:user][:id] = Base64.strict_encode64(session_params[:username])
+    credential_options[:user][:name] = session_params[:username]
+    credential_options[:user][:displayName] = session_params[:username]
+
+    credential_options[:challenge] = bin_to_str(credential_options[:challenge])
+    user.update!(current_challenge: credential_options[:challenge])
+
+    session[:username] = session_params[:username]
+
+    respond_to do |format|
+      format.json { render json: credential_options }
+    end
+  end
+
+  def callback
+    auth_response = WebAuthn::AuthenticatorAttestationResponse.new(
+      attestation_object: str_to_bin(params[:response][:attestationObject]),
+      client_data_json: str_to_bin(params[:response][:clientDataJSON])
+    )
+
+    user = User.find_by(username: session[:username])
+
+    raise "user #{session[:username]} never initiated sign up" unless user
+
+    render json: { status: "forbidden" },
+           status: :forbidden unless auth_response.valid?(str_to_bin(user.current_challenge), request.base_url)
+
+    if params[:response][:attestationObject].present?
```

Comment:
This conditional may be unnecessary after the split. I think we can now assume that this is always present on the Registration ceremony.

The original intention of that statement was to distinguish between registration vs authentication ceremonies.